Lib/weakref.py

```diff
@@ -384,11 +391,16 @@ def _commit_removals(self):
         # because a dead weakref never compares equal to a live weakref,
         # even if they happened to refer to equal objects.
         # However, it means keys may already have been removed.
-        l = self._pending_removals
+        pop = self._pending_removals.pop
         d = self.data
-        while l:
+        while True:
+            try:
+                key = pop()
+            except IndexError:
+                return
+
             try:
-                del d[l.pop()]
+                del d[key]
             except KeyError:
                 pass
```

Comment:
Observation: it kind of weirded me out there has to be a `KeyError` guard here. Even in the old code once you get something popped out of the list, you should be able to address the data key with it, no? There should be no duplicates in `self._pending_removals`!

Well, after some digging I found this isn't about duplicates. This is about a race where one `remove()` call hits the object while it's still `_iterating` (registering the key in `_pending_removals`), and concurrently another `remove()` call coming just a bit later when `_iterating` isn't set anymore (performing an eager removal).

In fact, while `_remove_dead_weakref` is called without a `try:except:` fence, it internally ignores raised KeyErrors. Similarly, the WeakSet is using `discard`.

There's some relevant info in https://bugs.python.org/issue21173 and https://bugs.python.org/issue28427.